cql3/expr/prepare_expr.cc

```diff
@@ -1057,6 +1058,182 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
+
+    if (oper == oper_t::IN) {
+        data_type rhs_receiver_type = list_type_impl::get_instance(std::move(lhs_type), false);
+        auto in_name = ::make_shared<column_identifier>(sstring("in(") + lhs_receiver->name->text() + sstring(")"), true);
+        return make_lw_shared<column_specification>(lhs_receiver->ks_name,
+                                                    lhs_receiver->cf_name,
+                                                    in_name,
+                                                    std::move(rhs_receiver_type));
+    } else if (oper == oper_t::CONTAINS) {
+        if (lhs_type->is_list()) {
+            return list_value_spec_of(*lhs_receiver);
+        } else if (lhs_type->is_set()) {
+            return set_value_spec_of(*lhs_receiver);
+        } else if (lhs_type->is_map()) {
+            return map_value_spec_of(*lhs_receiver);
+        } else {
+            throw exceptions::invalid_request_exception(format("Cannot use CONTAINS on non-collection column \"{}\"",
+                                                                     lhs_receiver->name));
+        }
+    } else if (oper == oper_t::CONTAINS_KEY) {
+        if (lhs_type->is_map()) {
+            return map_key_spec_of(*lhs_receiver);
+        } else {
+            throw exceptions::invalid_request_exception(format("Cannot use CONTAINS KEY on non-map column {}",
+                                                               lhs_receiver->name));
+        }
+    } else if (oper == oper_t::LIKE) {
+        if (!lhs_type->is_string()) {
+            throw exceptions::invalid_request_exception(
+                format("LIKE is allowed only on string types, which {} is not", lhs_receiver->name->text()));
+        }
+        return lhs_receiver;
+    } else {
+        return lhs_receiver;
+    }
+}
+
+static expression prepare_binop_rhs(const expression& rhs, const lw_shared_ptr<column_specification>& rhs_receiver, oper_t oper, data_dictionary::database db, const sstring& keyspace) {
+    // Lists of IN values must be handled separately during preparation
+    // because they might contain nulls and bind markers.
+    if (oper == oper_t::IN && is<collection_constructor>(rhs)) {
+        const collection_constructor& rhs_in_list = as<collection_constructor>(rhs);
+        return list_prepare_expression(rhs_in_list, db, keyspace, rhs_receiver, true);
+    }
+
+    return prepare_expression(rhs, db, keyspace, rhs_receiver);
+}
+
+binary_operator prepare_binary_operator(const binary_operator& binop, data_dictionary::database db, schema_ptr schema, prepare_context& ctx) {
+    expression prepared_lhs = prepare_binop_lhs(binop.lhs, db, *schema, binop);
+    lw_shared_ptr<column_specification> lhs_receiver = get_lhs_receiver(prepared_lhs, *schema);
+
+    lw_shared_ptr<column_specification> rhs_receiver = get_rhs_receiver(lhs_receiver, binop.op);
+    expression prepared_rhs = prepare_binop_rhs(binop.rhs, rhs_receiver, binop.op, db, schema->ks_name());
```

Comment:
This can certainly be a recursive call to prepare_expression().

I guess prepare_binop_lhs() can't be, because we don't have a receiver. This is the core of type inference here: lhs has to be an expression around a column, so we can determine the type, then infer the rhs type by looking at the lhs type and the operator. 

Later, we can make prepare_expression behave in two ways:
1. with a receiver - validate types against receiver
2. without receiver - return a receiver, or error if cannot be determined (e.g. bind variable)

Then binary operators will first try preparing without a receiver, both sides, and use the returned receiver to prepare the other side.

Reply:
`prepare_binop_rhs` is basically a wrapper of `prepare_expression`, but it does one more check to detect and deal with lists of IN values which can contain nulls.

Reply:
Let's see if we can move that check to runtime (and perhaps even: just ignore NULLs in IN lists, since no key can possibly match them).